### PR TITLE
fix(logql): Split ParserHint into ExtractionHints and LabelFilterHints

### PR DESCRIPTION
**Pull request by joe-elliott**
Splits the metric-extraction hints (`ExtractionHints`, built by `NewParserHint`) from the label-filter short-circuit (`LabelFilterHints`, built by the new `NewLabelFilterHints(stages)`) that were coupled in a single `ParserHint`/`Hints`. `BaseLabelsBuilder` now holds both independently, exposed via `ParserLabelHints()` and the new `LabelFilterHints()`. 

These two sets of functionality were distinct and pushing them together created complexity and resulted in at least one bug. For example in main `sum(rate({app="foo"}|logfmt|foo="bar"[5m]))` does not correctly apply the `foo="bar"` label filter hint due to a metrics related short circuit in `NewParserHint`.

### CODE REVIEW

File: pkg/logql/log/parser_hints.go [hunk 1]
```diff
@@ -213,3 +186,47 @@ func appendLabelHints(dst []string, src ...string) []string {
...
+			if len(f.RequiredLabelNames()) > 1 {
+				// Hints can only operate on one label at a time
+				continue
+			}
+			labelFilters = append(labelFilters, f)
+			labelNames = append(labelNames, f.RequiredLabelNames()...)
+		}
+	}
+	return &labelFilterHints{labelFilters: labelFilters, labelNames: labelNames}
+}
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply pkg/logql/log/parser_hints.go:232]: This is an existing bug and the one i was [actively working on](https://github.com/grafana/loki/pull/24415) when I decided that a split of these hints would help clarity. After this is merged I will rebase the linked PR, clean it up and fix this bug.

File: pkg/logql/log/parser_hints_test.go [hunk 2]
```diff
@@ -294,3 +294,28 @@ func TestLabelFiltersInParseHints(t *testing.T) {
 		require.True(t, h.ShouldContinueParsingLine("protocol", lb))
 	})
 }
+
+// Regression test: NewParserHint computes labelFilters/labelNames from stages
+// on every call, but 2 of its 4 early-return branches silently drop them
+// before returning, losing the ShouldContinueParsingLine short-circuit
+// entirely -- even for a plain filter with no CanModifyLabels concerns of
+// its own. A logfmt parser followed by a `foo="bar"` filter should always be
+// collectible into the hints, regardless of grouping/noLabels shape.
```
Comment (pracucci) [inline pkg/logql/log/parser_hints_test.go:303]: I think code comments should not talk about old version of code that doesn't exist anymore. They should just describe the current state.

File: pkg/logql/log/parser_hints_test.go [hunk 3]
```diff
@@ -289,8 +289,35 @@ func TestLabelFiltersInParseHints(t *testing.T) {
...
+	stages := []log.Stage{
+		log.NewLogfmtParser(false, false),
+		log.NewStringLabelFilter(labels.MustNewMatcher(labels.MatchEqual, "foo", "bar")),
+	}
+	labelFilterHints := log.NewLabelFilterHints(stages)
+	// foo=baz never matches the foo="bar" filter above.
+	lb := log.NewBaseLabelsBuilder().ForLabels(labels.FromStrings("app", "foo", "foo", "baz"), 0)
+
+	t.Run("noLabels with no required-label hints, e.g. sum(rate({app=\"foo\"}|logfmt|foo=\"bar\"[5m]))", func(t *testing.T) {
+		log.NewParserHint(nil, nil, false, true, "")
```
Comment (pracucci) [inline pkg/logql/log/parser_hints_test.go:315]: This return value is ignored. What is this call used for?

I think the whole test needs to be redesigned.

_Same comment applies below._

File: pkg/logql/log/parser_hints.go [hunk 4]
```diff
@@ -213,3 +186,47 @@ func appendLabelHints(dst []string, src ...string) []string {
...
+// NewLabelFilterHints scans stages for single-label filters so a parser can stop
+// extracting a line as soon as one of them fails to match a label it just extracted.
+func NewLabelFilterHints(stages []Stage) LabelFilterHints {
+	var labelNames []string
+	var labelFilters []LabelFilterer
+	for _, s := range stages {
+		switch f := s.(type) {
+		case *BinaryLabelFilter:
+			// TODO: as long as each leg of the binary filter operates on the same (and only 1) label,
+			// we should be able to add this to our filters
```
Comment (pracucci) [inline pkg/logql/log/parser_hints.go:220]: [non-blocking] I don't see much value in TODO comments in the code. They get persistently ignored. If it's an optimization worth pursuing, it should be an issue. Otherwise, I would just remove it and move on. If it's a good idea it will pop up in the future anyway.

File: pkg/logql/log/parser_hints.go [hunk 5]
```diff
@@ -213,3 +186,47 @@ func appendLabelHints(dst []string, src ...string) []string {
...
 	return dst
 }
+
+// NoLabelFilterHints returns a LabelFilterHints that never short-circuits.
+func NoLabelFilterHints() LabelFilterHints {
+	return &labelFilterHints{}
+}
+
+type labelFilterHints struct {
+	// Save the names next to the filters to avoid an alloc when f.RequiredLabelNames() is called
```
Comment (pracucci) [inline pkg/logql/log/parser_hints.go:196]: [nit] This comment was copied from the previous location, but I don't get it. Do you?

**Review by pracucci — COMMENTED** [review]
I think this PR accidentally introduce a new regression. See below:

```
sum by (bar) (rate({app="foo"} | logfmt | label_format foo="bar" | foo="bar" [5m]))
sum by (bar) (count_over_time({app="foo"} | logfmt | drop foo | foo="" [5m]))
sum by (bar) (count_over_time({app="foo"} | logfmt | keep bar | foo="" [5m]))
```

All three should yield a sample of 1 for the line `foo=nope bar=hello`, but return none on this branch.

I'm OK if you're going to fix it as part of https://github.com/grafana/loki/pull/24415 but we need to make sure the two PRs land within the same week, otherwise we're introducing a regression.

File: pkg/logql/log/parser_hints_test.go (hunk 2 above)
Comment (joe-elliott) [reply pkg/logql/log/parser_hints_test.go]: agreed and apologize for this. i sometimes let claude get away from me on the comments.

removed in https://github.com/grafana/loki/pull/24565/changes/220c2e820e3a2e3f0472cb113f3f2c29e3f7b3f7

File: pkg/logql/log/parser_hints.go (hunk 4 above)
Comment (joe-elliott) [reply pkg/logql/log/parser_hints.go]: fixed in https://github.com/grafana/loki/pull/24565/changes/220c2e820e3a2e3f0472cb113f3f2c29e3f7b3f7

the label hints parser now supports `BinaryLabelFilter`s that only reference one label.

File: pkg/logql/log/parser_hints.go (hunk 5 above)
Comment (joe-elliott) [reply pkg/logql/log/parser_hints.go]: it's saying that when we [evalute the hint](https://github.com/grafana/loki/blob/3105b2b19e04865de7f424f2a3ec67a29975d6c3/pkg/logql/log/parser_hints.go#L134-L142) we could call [RequiredLabelNames](https://github.com/grafana/loki/blob/3105b2b19e04865de7f424f2a3ec67a29975d6c3/pkg/logql/log/pipeline.go#L40) on each LabelFilterer to determine if it applied, but this may cause allocs. So instead we're memoizing the work in the `labelNames` slice.

File: pkg/logql/log/parser_hints_test.go (hunk 3 above)
Comment (joe-elliott) [reply pkg/logql/log/parser_hints_test.go]: test [removed](https://github.com/grafana/loki/pull/24565/commits/220c2e820e3a2e3f0472cb113f3f2c29e3f7b3f7) and replaced with two tests:

`TestNewLabelFilterHints_Boundary` that passes various LogQL shapes into the label hints simply to confirm which ones the label hints applies to.

`TestNewLabelFilterHintsMatchers` that passes the known good shapes to the hint and tests `ShouldContinueParsingLine` for expected results

File: pkg/logql/internal/logqltest/testdata/label_filter_hints.logqltest [hunk 6]
```diff
@@ -0,0 +1,191 @@
+# LabelFilterHints (pkg/logql/log/parser_hints.go) lets a parser stop
```
Comment (joe-elliott) [inline pkg/logql/internal/logqltest/testdata/label_filter_hints.logqltest]: i made a standalone file to focus on labelfilterhints edge cases. unsure if that is the pattern we want for logql DSL tests. open to suggestions

**Comment by joe-elliott** [comment]
Ready for review @pracucci. cc @benclive

File: pkg/logql/log/parser_hints.go [hunk 7]
```diff
@@ -213,3 +186,96 @@ func appendLabelHints(dst []string, src ...string) []string {
...
+	return true
+}
+
+// NewLabelFilterHints scans stages for label filters so a parser can stop
+// extracting a line as soon as one of them fails to match a label it just extracted.
+func NewLabelFilterHints(stages []Stage) LabelFilterHints {
+	var labelNames []string
+	var labelFilters []LabelFilterer
+
+	isParser := func(s Stage) bool {
```
Comment (joe-elliott) [inline pkg/logql/log/parser_hints.go]: possibly belongs as a function in parser.go? other implementation might be a common interface they all implement with a "seal trait" like:

https://github.com/grafana/loki/blob/main/pkg/logql/log/label_filter.go#L64
